src/parser/jobs/mnk/BuffUptime.js

```diff
@@ -0,0 +1,161 @@
+import React, {Fragment} from 'react'
+
+import {ActionLink} from 'components/ui/DbLink'
+import ACTIONS, {getAction} from 'data/ACTIONS'
+import STATUSES from 'data/STATUSES'
+import Module from 'parser/core/Module'
+import {Rule, Requirement} from 'parser/core/modules/Checklist'
+import {Suggestion, SEVERITY} from 'parser/core/modules/Suggestions'
+
+const GCD_CYCLE_LENGTH = 5
+
+export default class BuffUptime extends Module {
+	static handle = 'BuffUptime'
+	static dependencies = [
+		'checklist',
+		'combatants',
+		'enemies',
+		'invuln',
+		'suggestions',
+	]
+
+	// Dragon kick can be applied to multiple enemies
+	_lastDragonKickUse = {}
+	_lastTwinSnakesUse = null
+
+	_earlyDragonKicks = 0
+	_earlyTwinSnakes = 0
+
+	_gcdsSinceDK = {}
+	_gcdsSinceTS = 0
+
+	constructor(...args) {
+		super(...args)
+
+		// Hook all GCDs so we can count GCDs in buff windows
+		this.addHook('cast', {by: 'player'}, this._onCast)
+
+		// Apparently DK is a bit busted but let's use the debuff for now
+		this.addHook(['applybuff', 'refreshbuff'], {by: 'player', abilityId: STATUSES.BLUNT_RESISTANCE_DOWN.id}, this._onDragonKick)
+		this.addHook(['applybuff', 'refreshbuff'], {by: 'player', abilityId: STATUSES.TWIN_SNAKES.id}, this._onTwinSnakes)
+
+		this.addHook('complete', this._onComplete)
+	}
+
+	_onCast(event) {
+		const action = getAction(event.ability.guid)
+
+		if (!action) {
+			return
+		}
+
+		if (action.onGcd) {
+			this._gcdsSinceDK[event.targetID] += 1
+			this._gcdsSinceTS++
+		}
+	}
+
+	_onDragonKick(event) {
+		// Make sure we're tracking for this target
+		const lastApplication = this._lastDragonKickUse[event.targetID]
+
+		// If it's not been applied yet set it and skip out
+		if (!lastApplication) {
+			this._lastDragonKickUse[event.targetID] = event.timestamp
+			return
+		}
+
+		// This logic is busted if the player's GCD is under 1.67s
```

Comment:
_Enhanced Arrow PTSD flashbacks._ :internalscreaming:

Reply:
Could always check the estimate etc, but that's a lot of logic for little gain.